### PR TITLE
MM-52655 Allow plugin requests to include Authorization headers from external systems

### PR DESCRIPTION
**Pull request by kyeongsoosoo**
#### Summary

The `Authorization` header is stripped from any request going to a plugin.
I changed to remove the `Authorization` header only if the request is from an authenticated user.

#### Ticket Link

Fixes https://github.com/mattermost/mattermost/issues/23575
Jira https://mattermost.atlassian.net/browse/MM-52655

#### Release Note

```release-note
Allow plugin requests to include Authorization headers from external systems
```

### CODE REVIEW

File: server/channels/app/plugin_requests.go [hunk 1]
```diff
@@ -208,6 +208,9 @@ func (ch *Channels) servePluginRequest(w http.ResponseWriter, r *http.Request, h
 		if (session != nil && session.Id != "") && err == nil && csrfCheckPassed {
 			r.Header.Set("Mattermost-User-Id", session.UserId)
 			context.SessionId = session.Id
+
+			r.Header.Del(model.HeaderAuth)
+
 		}
```
Comment (mickmister) [inline server/channels/app/plugin_requests.go:214]: Nit: This empty newline is not needed below `r.Header.Del`. Wondering why the linter didn't catch this, maybe we do this in some places

```suggestion
			context.SessionId = session.Id

			r.Header.Del(model.HeaderAuth)
		}
```

**Review by mickmister — APPROVED** [review]
LGTM :+1:  Just one non-blocking suggestion for code styling

Thanks for the contribution @kyeongsoosoo!

File: server/channels/app/plugin_requests.go (hunk 1 above)
Comment (kyeongsoosoo) [reply server/channels/app/plugin_requests.go]: I fixed the code following the suggestion!

**Review by mickmister — APPROVED** [review]

**Review by hanzei — APPROVED** [review]
LGTM

**Comment by hanzei** [comment]
@esarafianou Would you please also give these changes a review?

**Comment by hanzei** [comment]
@esarafianou Plugins act as an HTTP server. They serve requests from MM clients (which are authenticated in a Mattermost sense) but might also serve HTTP requests from external services like webhooks. These requests from external services might use the `Authorization` header to authorize themselves against the plugin.

Does that make things more clear?

**Comment by esarafianou** [comment]
@hanzei thanks for the explanation. The main thing I'm worried about is accidentally logging the Authorization header in the plugins. Nothing related to this PR, but something to keep in mind when designing plugins that don't require a Mattermost authenticated session. Consider this PR approved by me.

**Comment by hanzei** [comment]
@saturninoabril Any idea why the e2e test timed out?

**Comment by saturninoabril** [comment]
I'm not sure. I did restart and see that the service for webhook testing is infinitely waiting. I'll update this branch and will trigger the test again.

**Comment by hanzei** [comment]
Thanks @saturninoabril :+1:

**Comment by hanzei** [comment]
@DHaussermann Do you want to validate the changes by setting up a plugin that listens to incoming HTTP requests with an `Authorization` header?

**Comment by mickmister** [comment]
@kyeongsoosoo Would you be interested in implementing a feature in the [demo plugin](https://github.com/mattermost/mattermost-plugin-demo) that tests this new functionality? No worries if not 👍

**Comment by kyeongsoosoo** [comment]
@mickmister I'll try it this week😀

**Comment by kyeongsoosoo** [comment]
> @kyeongsoosoo Would you be interested in implementing a feature in the [demo plugin](https://github.com/mattermost/mattermost-plugin-demo) that tests this new functionality? No worries if not 👍

@mickmister  I created a PR for this!

**Comment by hanzei** [comment]
@DHaussermann This PR and https://github.com/mattermost/mattermost-plugin-demo/pull/170 are ready for your review

**Comment by hanzei** [comment]
@esarafianou Would you mind giving this PR a review?

**Comment by hanzei** [comment]
@esarafianou Scratch that, I forgot that we already talked about this PR

**Comment by mickmister** [comment]
@DHaussermann Maybe @AayushChaudhary0001 can help with testing this?

**Comment by hanzei** [comment]
@mickmister Given that https://github.com/mattermost/mattermost-plugin-demo/pull/170 got merged, can we also merge this PR?

**Comment by mickmister** [comment]
Thanks for the contribution @kyeongsoosoo! 🚀

**Comment by cwarnermm** [comment]
@hanzei @mickmister - Would either of you be open to drafting a developer documentation update that addresses the functionality change introduced via this PR?

**Comment by mickmister** [comment]
@cwarnermm There's a detailed explanation in the linked ticket https://mattermost.atlassian.net/browse/MM-52655

Though a summary of the change would be:

> HTTP requests to server-side plugins can now use an `Authorization` header in the request for the plugin to use, as long as the token provided in the header is not a user token issued by the Mattermost server. This is useful for connecting external systems that require authentication through an `Authorization` header with their own token.

**Pull request merged**